agent/engine/task_manager.go

```diff
@@ -481,6 +481,45 @@ func (mtask *managedTask) handleContainerChange(containerChange dockerContainerC
 	// Container has progressed its status if we reach here. Make sure to save it to database.
 	defer mtask.engine.saveContainerData(container)
 
+	// If container is transitioning to STOPPED, first check if we should short-circuit
+	// the stop workflow and restart the container.
+	if event.Status == apicontainerstatus.ContainerStopped && container.RestartPolicyEnabled() {
+		exitCode := event.DockerContainerMetadata.ExitCode
+		shouldRestart, reason := container.RestartTracker.ShouldRestart(exitCode, container.GetStartedAt(), container.GetDesiredStatus())
+		if shouldRestart {
+			defer container.RestartTracker.RecordRestart()
```

Comment:
Could you explain what happens when a container restart fails? Since `LastRestartAt` is updated even though the container didn't restart, is the next restart attempt delayed?

Reply:
If the "StartContainer" call itself fails, then the container will never be attempted to be restarted again. The restart policy is only triggered on containers that have moved past the "StartContainer" phase into RUNNING. The container also needs to be in the RUNNING state for a minimum time for it to be eligible to be restarted (this is the [RestartAttemptPeriod](https://pkg.go.dev/github.com/aws/amazon-ecs-agent/ecs-agent@v0.0.0-20240509232427-1133dd999145/api/container/restart#RestartPolicy) field)